ext/standard/array.c

```diff
@@ -3446,34 +3603,57 @@ PHP_FUNCTION(array_slice)
 	// Contains modified variants of ZEND_HASH_FOREACH_VAL
 	{
 		HashTable *ht = Z_ARRVAL_P(input);
-		Bucket *p = find_bucket_at_offset(ht, offset);
-		Bucket *end = ht->arData + ht->nNumUsed;
 
 		/* Start at the beginning and go until we hit offset */
-		if (HT_IS_PACKED(Z_ARRVAL_P(input)) &&
-			(!preserve_keys ||
-			 (offset == 0 && HT_IS_WITHOUT_HOLES(Z_ARRVAL_P(input))))) {
-
-			zend_hash_real_init_packed(Z_ARRVAL_P(return_value));
-			ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(return_value)) {
-				for (; p != end; p++) {
-					if (__fill_idx >= length) {
-						break;
+		if (HT_IS_PACKED(ht)) {
+			zval *zv = find_packed_val_at_offset(ht, offset);
+			zval *end = ht->arPacked + ht->nNumUsed;
+
+			if (!preserve_keys
+			 || (offset == 0 && HT_IS_WITHOUT_HOLES(ht))) {
+				zend_hash_real_init_packed(Z_ARRVAL_P(return_value));
+				ZEND_HASH_FILL_PACKED(Z_ARRVAL_P(return_value)) {
+					for (; zv != end; zv++) {
+						if (__fill_idx >= length) {
+							break;
+						}
+						if (UNEXPECTED(Z_TYPE_P(zv) == IS_UNDEF)) {
```

Comment:
This check looks unnecessary, as the ht is WITHOUT_HOLES.

Reply:
No. WITHIUT_HOLES is checked only if ``preserve_keys`` is true. Removing this, breaks ext/standard/tests/array/array_slice_variation11.phpt.